src/socket/udp_socket.cr

```diff
@@ -92,4 +92,154 @@ class UDPSocket < IPSocket
+  def multicast_interface(address : IPAddress)
+    if @family == Family::INET
+      if addr = address.@addr4
+        setsockopt LibC::IP_MULTICAST_IF, addr, LibC::IPPROTO_IP
+      else
+        raise Socket::Error.new "Expecting an IPv4 interface address. Address provided: #{address.address}"
+      end
+    else
+      raise Socket::Error.new "Unsupported IP address family: #{@family}. For use with IPv4 only"
+    end
+  end
+
+  # For hosts with multiple interfaces, each multicast transmission is sent
+  # from the primary network interface. This function overrides the default
+  # IPv6 interface for subsequent transmissions. Setting the interface to
+  # index `0` will select the default interface.
+  def multicast_interface(index : UInt32)
+    if @family == Family::INET6
+      setsockopt LibC::IPV6_MULTICAST_IF, index, LibC::IPPROTO_IPV6
+    else
+      raise Socket::Error.new "Unsupported IP address family: #{@family}. For use with IPv6 only"
+    end
+  end
+
+  # A host must become a member of a multicast group before it can receive
+  # datagrams sent to the group.
+  # Raises `Socket::Error` if an incompatible address is provided.
+  def join_group(address : IPAddress)
+    case @family
+    when Family::INET
+      group_modify(address, LibC::IP_ADD_MEMBERSHIP)
+    when Family::INET6
+      group_modify(address, LibC::IPV6_JOIN_GROUP)
+    else
+      raise Socket::Error.new "Unsupported IP address family: #{@family}"
+    end
+  end
+
+  # Drops membership to the specified group. Memberships are automatically
+  # dropped when the socket is closed or the process exits.
+  # Raises `Socket::Error` if an incompatible address is provided.
+  def leave_group(address : IPAddress)
+    case @family
+    when Family::INET
+      group_modify(address, LibC::IP_DROP_MEMBERSHIP)
+    when Family::INET6
+      group_modify(address, LibC::IPV6_LEAVE_GROUP)
+    else
+      raise Socket::Error.new "Unsupported IP address family: #{@family}"
+    end
+  end
+
+  private def group_modify(ip, operation)
+    resp = case @family
+           when Family::INET
+             if ip_addr = ip.@addr4
+               req = LibC::IpMreq.new
+               req.imr_multiaddr = ip_addr
+
+               setsockopt operation, req, LibC::IPPROTO_IP
+             else
+               raise Socket::Error.new "Expecting an IPv4 multicast address. Address provided: #{ip.address}"
+             end
+           when Family::INET6
+             if ip_addr = ip.@addr6
+               req = LibC::Ipv6Mreq.new
+               req.ipv6mr_multiaddr = ip_addr
+
+               setsockopt operation, req, LibC::IPPROTO_IPV6
+             else
+               raise Socket::Error.new "Expecting an IPv6 multicast address. Address provided: #{ip.address}"
+             end
+           else
+             raise Socket::Error.new "Unsupported IP address family: #{@family}"
+           end
+    # FIXME: We yield so that when we perform a send we'll receive the loopback if
```

Comment:
Was this discussed at some point? What code fails without this? Maybe there is a way to avoid `Fiber.yield` here.

(but it can be discussed in a separate issue if everything is working fine)

Reply:
@asterite in my testing I found that without the yield the following wouldn't work as expected:

```crystal
# Loop back indicates we want to receive anything we send back on the same socket
udp.multicast_loopback = true
udp.join_group(addr) # (yielding here)
udp.send("testing", addr)

# without the yield when joining the group above then the multicast send above succeeds but is not received locally even though loopback is enabled
udp.receive[0].should eq("testing")
```

The processes handling multicast in the background seem to be quite complicated so I assume it's related to some kernel data-structures but I am not certain
https://medium.com/tcp-ip/igmp-query-group-join-leave-681ee7a00db1
https://www.eventhelix.com/RealtimeMantra/Networking/igmp/igmp-query-join-leave.pdf

Reply:
Okay, so what `Fiber.yield` does is to put the current fiber to sleep and see if another fiber can execute. If that's the case the other fiber executes until it gets blocked, then other fibers (like the first one) resume execution.

I guess what I'm not seeing here is is that in your code there's just a single fiber, so `Fiber.yield` would have no effect, except maybe waiting a few (milli/nano)seconds to resume execution. That's why I think that `Fiber.yield` there is a big smell. There's probably some misunderstanding around how this behaves and it's probably not good to introduce something that we don't understand.

Reply:
So remove the fiber yield?
My feeling is that the loopback isn't enabled until the IGMP join group packet is transmitted.

So the tests might pass if I remove this yield and set `multicast_hops = 0` so that the packet send isn't required.